File: dotnet/src/MableBank.Application/ProcessDay.cs
using MableBank.Application.Telemetry;
using MableBank.Domain;

namespace MableBank.Application;

/// <summary>Orchestrates loading balances, processing transfers, observability, and reconciliation.</summary>
public sealed class ProcessDay
{
    private readonly IAccountBalanceReader _balancesReader;
    private readonly ITransferInstructionReader _transfersReader;
    private readonly bool _dryRun;
    private readonly ILogger _logger;
    private readonly IMetrics _metrics;

    public ProcessDay(
        IAccountBalanceReader balancesReader,
        ITransferInstructionReader transfersReader,
        bool dryRun = false,
        ILogger? logger = null,
        IMetrics? metrics = null)
    {
        _balancesReader = balancesReader ?? throw new ArgumentNullException(nameof(balancesReader));
        _transfersReader = transfersReader ?? throw new ArgumentNullException(nameof(transfersReader));
        _dryRun = dryRun;
        _logger = logger ?? NullLogger.Instance;
        _metrics = metrics ?? NullMetrics.Instance;
    }

    /// <summary>Runs the processing day and returns a report of outcomes and final balances.</summary>
    public Task<ProcessingReport> CallAsync(CancellationToken cancellationToken = default) =>
        CallAsyncCore(cancellationToken);

    private async Task<ProcessingReport> CallAsyncCore(CancellationToken cancellationToken)
    {
        var runId = Guid.NewGuid();
        var accountBook = await LoadAccountBalances.FromReaderAsync(_balancesReader, cancellationToken)
            .ConfigureAwait(false);
        var journal = new InMemoryTransferJournal();
        IIdempotencyRegistry? registry = _dryRun ? null : new InMemoryIdempotencyRegistry();
        var processor = new ProcessTransfers(
            accountBook,
            _dryRun,
            registry,
            journal,
            runId);
        var recorder = new TransferResultRecorder();

        await foreach (var instruction in _transfersReader
                           .ReadInstructionsAsync(cancellationToken)
                           .ConfigureAwait(false))
        {
            var result = recorder.Record(processor.Process(instruction));
            ObserveTransfer(result);
        }

        if (!_dryRun)
        {
            Reconciliation.Verify(accountBook);
        }

        return new ProcessingReport(recorder, accountBook, _dryRun, journal, runId);
    }

    private void ObserveTransfer(TransferResult result)
    {
        var telemetry = TransferProcessedTelemetry.FromResult(result);
        _metrics.RecordTransferProcessed(telemetry);
        _logger.LogTransferProcessed(telemetry);
    }
}

File: nodejs-typescript/src/application/ProcessDay.ts
import { verifyReconciliation } from '../domain/Reconciliation.js';
import type { TransferResult } from '../domain/TransferResult.js';
import { TransferProcessedTelemetry } from './telemetry/TransferProcessedTelemetry.js';
import type { AccountBalanceReader, TransferInstructionReader } from './ports/readers.js';
import { loadAccountBalances } from './LoadAccountBalances.js';
import type { Logger } from './Logger.js';
import type { Metrics } from './Metrics.js';
import { NullLogger } from './NullLogger.js';
import { NullMetrics } from './NullMetrics.js';
import { ProcessingReport } from './ProcessingReport.js';
import { ProcessTransfers, newRunId } from './ProcessTransfers.js';
import { TransferResultRecorder } from './TransferResultRecorder.js';
import { InMemoryIdempotencyRegistry } from './IdempotencyRegistry.js';
import { InMemoryTransferJournal } from './TransferJournal.js';

export type { AccountBalanceReader, TransferInstructionReader } from './ports/readers.js';

export type ProcessDayOptions = {
  readonly balancesReader: AccountBalanceReader;
  readonly transfersReader: TransferInstructionReader;
  readonly dryRun?: boolean;
  readonly logger?: Logger;
  readonly metrics?: Metrics;
};

export class ProcessDay {
  private readonly logger: Logger;
  private readonly metrics: Metrics;

  constructor(private readonly options: ProcessDayOptions) {
    this.logger = options.logger ?? new NullLogger();
    this.metrics = options.metrics ?? new NullMetrics();
  }

  call(): ProcessingReport {
    const runId = newRunId();
    const accountBook = loadAccountBalances(this.options.balancesReader);
    const journal = new InMemoryTransferJournal();
    const registry = this.options.dryRun ? null : new InMemoryIdempotencyRegistry();
    const processor = new ProcessTransfers(
      accountBook,
      this.options.dryRun ?? false,
      registry,
      journal,
      runId,
    );
    const recorder = new TransferResultRecorder();

    this.options.transfersReader.eachInstruction((instruction) => {
      const result = recorder.record(processor.process(instruction));
      this.observeTransfer(result);
    });

    if (!this.options.dryRun) {
      verifyReconciliation(accountBook);
    }

    return new ProcessingReport(recorder, accountBook, this.options.dryRun ?? false, journal, runId);
  }

  private observeTransfer(result: TransferResult): void {
    const telemetry = TransferProcessedTelemetry.fromResult(result);
    this.metrics.recordTransferProcessed(telemetry);
    this.logger.logTransferProcessed(telemetry);
  }
}
